File: apk/app/src/main/java/com/example/momentra/ui/shell/group/shared/GroupTabDataCache.kt
package com.example.momentra.ui.shell.group.shared

import com.example.momentra.data.api.ActivityItemDto
import com.example.momentra.data.api.AnalyticsInsightItemDto
import com.example.momentra.data.api.GroupFinancePayloadDto
import com.example.momentra.data.api.GroupLifePayloadDto
import com.example.momentra.data.api.GroupMemoryPayloadDto
import com.example.momentra.data.api.GroupParticipantDto
import com.example.momentra.data.api.GroupPulsePayloadDto
import java.util.concurrent.ConcurrentHashMap

/** In-memory SWR cache for Group tab datasets keyed by momentId. */
object GroupTabDataCache {
    data class PulseTab(
        val title: String?,
        val pulse: GroupPulsePayloadDto?,
        val finance: GroupFinancePayloadDto?,
        val activities: List<ActivityItemDto>,
        val insights: List<AnalyticsInsightItemDto> = emptyList(),
    )

    data class MemoryTab(
        val memory: GroupMemoryPayloadDto?,
        val finance: GroupFinancePayloadDto?,
        val pulse: GroupPulsePayloadDto?,
        val participants: List<GroupParticipantDto> = emptyList(),
    )

    private val pulseByMoment = ConcurrentHashMap<String, PulseTab>()
    private val memoryByMoment = ConcurrentHashMap<String, MemoryTab>()
    private val lifeByMoment = ConcurrentHashMap<String, GroupLifePayloadDto>()

    fun peekPulse(momentId: String): PulseTab? = pulseByMoment[momentId]

    fun putPulse(momentId: String, data: GroupPulseTabData) {
        pulseByMoment[momentId] = PulseTab(
            title = data.title,
            pulse = data.pulse,
            finance = data.finance,
            activities = data.activities,
            insights = data.insights,
        )
    }

    fun peekMemory(momentId: String): MemoryTab? = memoryByMoment[momentId]

    fun putMemory(momentId: String, data: MemoryTab) {
        memoryByMoment[momentId] = data
    }

    fun peekLife(momentId: String): GroupLifePayloadDto? = lifeByMoment[momentId]

    fun putLife(momentId: String, payload: GroupLifePayloadDto?) {
        if (payload != null) {
            lifeByMoment[momentId] = payload
        }
    }

    /** Drop cached facets for one moment after a write so SWR reloads fresh data. */
    fun invalidateMoment(momentId: String) {
        pulseByMoment.remove(momentId)
        memoryByMoment.remove(momentId)
        lifeByMoment.remove(momentId)
    }

    fun clear() {
        pulseByMoment.clear()
        memoryByMoment.clear()
        lifeByMoment.clear()
    }
}

data class GroupPulseTabData(
    val title: String?,
    val pulse: GroupPulsePayloadDto?,
    val finance: GroupFinancePayloadDto?,
    val activities: List<ActivityItemDto>,
    val insights: List<AnalyticsInsightItemDto> = emptyList(),
)

File: momentra/momentra/Shell/GroupActive/Shared/GroupTabDataCache.swift
import Foundation

/// In-memory SWR cache for Group tab datasets keyed by momentId.
enum GroupTabDataCache {
    struct PulseTab {
        let title: String?
        let pulse: APIClient.GroupPulsePayload?
        let finance: APIClient.GroupFinancePayload?
        let activities: [APIClient.ActivityItemPayload]
        var insights: [AnalyticsInsightItemPayload]

        init(
            title: String?,
            pulse: APIClient.GroupPulsePayload?,
            finance: APIClient.GroupFinancePayload?,
            activities: [APIClient.ActivityItemPayload],
            insights: [AnalyticsInsightItemPayload] = []
        ) {
            self.title = title
            self.pulse = pulse
            self.finance = finance
            self.activities = activities
            self.insights = insights
        }
    }

    struct MemoryTab {
        let memory: APIClient.GroupMemoryPayload?
        let finance: APIClient.GroupFinancePayload?
        let pulse: APIClient.GroupPulsePayload?
        var participants: [APIClient.GroupParticipantPayload] = []
    }

    private static var pulseByMoment: [String: PulseTab] = [:]
    private static var memoryByMoment: [String: MemoryTab] = [:]
    private static var lifeByMoment: [String: APIClient.GroupLifePayload] = [:]

    static func peekPulse(_ momentId: String) -> PulseTab? {
        pulseByMoment[momentId]
    }

    static func putPulse(_ momentId: String, _ tab: PulseTab) {
        pulseByMoment[momentId] = tab
    }

    static func peekMemory(_ momentId: String) -> MemoryTab? {
        memoryByMoment[momentId]
    }

    static func putMemory(_ momentId: String, _ tab: MemoryTab) {
        memoryByMoment[momentId] = tab
    }

    static func peekLife(_ momentId: String) -> APIClient.GroupLifePayload? {
        lifeByMoment[momentId]
    }

    static func putLife(_ momentId: String, _ payload: APIClient.GroupLifePayload) {
        lifeByMoment[momentId] = payload
    }

    /// Drop cached facets for one moment after a write so SWR reloads fresh data.
    static func invalidateMoment(_ momentId: String) {
        pulseByMoment.removeValue(forKey: momentId)
        memoryByMoment.removeValue(forKey: momentId)
        lifeByMoment.removeValue(forKey: momentId)
    }

    static func clear() {
        pulseByMoment.removeAll()
        memoryByMoment.removeAll()
        lifeByMoment.removeAll()
    }
}

enum GroupTabPrefetch {
    static let activityLimit = 5

    /// Warm critical-path pulse+finance+activity (inflight-deduped via GroupTabLoad).
    static func run(momentId: String) async {
        guard !momentId.isEmpty else { return }
        _ = try? await GroupTabLoad.loadPulseTab(momentId: momentId)
    }
}
